# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/QuotaManagerImpl.h
# Commit: d63dac1ee4d4
# Full Hash: d63dac1ee4d471defff484dae5a3535e1f8ba072
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-10-10 22:03:45
# Regressor Bug: 1866402
# File Overlap Count: 1
# Description:
#   Bug 1866402 - Make it possible to initialize temporary origins without ensuring origin directories; r=dom-storage-reviewers,jari
#   
#   LSNG already uses some QuotaManager APIs to achieve that origin directories are
#   not created if they don't exist during datastore preparation, but the feature
#   is not easy to use and it's also not generalized enough for use in other quota
# ==============================================================================

diff -r 2ab08c49f2b1 -r d63dac1ee4d4 dom/quota/QuotaManagerImpl.h
--- a/dom/quota/QuotaManagerImpl.h	Thu Oct 10 12:22:54 2024 +0000
+++ b/dom/quota/QuotaManagerImpl.h	Thu Oct 10 12:22:55 2024 +0000
@@ -14,6 +14,19 @@
 
 namespace mozilla::dom::quota {
 
+template <typename F>
+auto QuotaManager::WithOriginInfo(const OriginMetadata& aOriginMetadata,
+                                  F aFunction)
+    -> std::invoke_result_t<F, const RefPtr<OriginInfo>&> {
+  MutexAutoLock lock(mQuotaMutex);
+
+  RefPtr<OriginInfo> originInfo =
+      LockedGetOriginInfo(aOriginMetadata.mPersistenceType, aOriginMetadata);
+  MOZ_ASSERT(originInfo);
+
+  return aFunction(originInfo);
+}
+
 template <typename P>
 void QuotaManager::CollectPendingOriginsForListing(P aPredicate) {
   MutexAutoLock lock(mQuotaMutex);
@@ -28,7 +41,7 @@
         pair->LockedGetGroupInfo(PERSISTENCE_TYPE_DEFAULT);
     if (groupInfo) {
       for (const auto& originInfo : groupInfo->mOriginInfos) {
-        if (!originInfo->mDirectoryExists) {
+        if (!originInfo->LockedDirectoryExists()) {
           aPredicate(originInfo);
         }
       }